lib/core/src/wallet.rs

```diff
@@ -206,7 +211,31 @@ impl OnchainWallet for LiquidOnchainWallet {
 
     /// Get the next unused address in the wallet
     async fn next_unused_address(&self) -> Result<Address, PaymentError> {
-        Ok(self.wallet.lock().await.address(None)?.address().clone())
+        let tip = self.tip().await.height();
+        let address = match self.persister.next_expired_reserved_address(tip)? {
+            Some(reserved_address) => {
+                debug!(
+                    "Got reserved address {} that expired on block height {}",
+                    reserved_address.address, reserved_address.expiry_block_height
+                );
+                ElementsAddress::from_str(&reserved_address.address)
+                    .map_err(|e| PaymentError::Generic { err: e.to_string() })?
+            }
+            None => {
+                let next_index = self.persister.next_derivation_index()?;
+                let address_result = self.wallet.lock().await.address(next_index)?;
+                let address = address_result.address().clone();
+                let index = address_result.index();
+                debug!(
+                    "Got unused address {} with derivation index {}",
+                    address, index
+                );
+                self.persister.set_last_derivation_index(index)?;
```

Comment:
It seems that we don't need this function as the next_derivation_index already increments and update the index.
I suggest to remove it so the only way to update the index is actually by incrementing it.

Reply:
@roeierez the cached last derivation index will never be set (`next_derivation_index()` will always return None) if we remove it. How about we only set it if `next_derivation_index()` returns None?

Reply:
I see, it is here for the migration case.
In that case shouldn't we set the index to index + 1 ? It seems that the persisted value should be the next available (not used) index.